File: R/R/palettes.R
#' List all available journal keys
#'
#' @return A character vector of journal keys, sorted alphabetically.
#' @export
#' @examples
#' list_journals()
list_journals <- function() {
  sort(names(.get_data()))
}

#' List all palette names for a journal
#'
#' @param journal Character. Journal key (case-insensitive).
#'   Use [list_journals()] to see available options.
#' @return A character vector of palette names, sorted alphabetically.
#' @export
#' @examples
#' list_palettes("nature")
list_palettes <- function(journal) {
  data <- .get_data()
  key <- tolower(journal)
  if (!key %in% names(data)) {
    stop(sprintf(
      "Journal '%s' not found. Available: %s",
      journal, paste(sort(names(data)), collapse = ", ")
    ))
  }
  sort(names(data[[key]][["palettes"]]))
}

#' Get full palette metadata for a journal
#'
#' @param journal Character. Journal key (case-insensitive).
#' @param palette Character. Palette name within that journal. Default `"main"`.
#' @return A list with elements `colors`, `colorblind_safe`, `description`, `type`.
#' @export
#' @examples
#' p <- get_palette("nature")
#' p$colors[1:3]
get_palette <- function(journal, palette = "main") {
  data <- .get_data()
  key <- tolower(journal)
  if (!key %in% names(data)) {
    stop(sprintf(
      "Journal '%s' not found. Available: %s",
      journal, paste(sort(names(data)), collapse = ", ")
    ))
  }
  palettes <- data[[key]][["palettes"]]
  pal <- tolower(palette)
  if (!pal %in% names(palettes)) {
    stop(sprintf(
      "Palette '%s' not found for '%s'. Available: %s",
      palette, journal, paste(sort(names(palettes)), collapse = ", ")
    ))
  }
  palettes[[pal]]
}

#' Get hex color codes for a journal palette
#'
#' @param journal Character. Journal key (case-insensitive).
#' @param palette Character. Palette name. Default `"main"`.
#' @param n Integer or NULL. Number of colors to return. If NULL, returns all.
#'   If `n` exceeds the palette length, colors are cycled.
#' @param colorblind_only Logical. If TRUE, error when palette is not colorblind-safe.
#' @return A character vector of hex color codes.
#' @export
#' @examples
#' get_colors("nature", n = 3)
#' get_colors("colorblind", "okabe_ito", colorblind_only = TRUE)
get_colors <- function(journal, palette = "main", n = NULL, colorblind_only = FALSE) {
  pal <- get_palette(journal, palette)

  if (colorblind_only && !isTRUE(pal[["colorblind_safe"]])) {
    stop(sprintf(
      "Palette '%s' for '%s' is not colorblind-safe. ",
      palette, journal
    ))
  }

  colors <- unlist(pal[["colors"]])

  if (is.null(n)) {
    return(colors)
  }

  n <- as.integer(n)
  if (n <= length(colors)) {
    return(colors[seq_len(n)])
  }

  # Cycle colors if n > palette length
  rep_colors <- rep(colors, times = ceiling(n / length(colors)))
  rep_colors[seq_len(n)]
}

File: src/pubchroma/palettes.py
"""Core palette loading and recommendation functions."""

from __future__ import annotations

import json
from importlib.resources import files
from typing import Optional


def _load_data() -> dict:
    """Load palette data from the bundled JSON file."""
    data_path = files("pubchroma").joinpath("data/journals.json")
    return json.loads(data_path.read_text(encoding="utf-8"))


_DATA: dict | None = None


def _get_data() -> dict:
    global _DATA
    if _DATA is None:
        _DATA = _load_data()
    return _DATA


def list_journals() -> list[str]:
    """Return all available journal keys.

    Returns
    -------
    list[str]
        Sorted list of journal keys (e.g. ['bmj', 'cell', 'jama', ...]).

    Examples
    --------
    >>> import pubchroma as pc
    >>> pc.list_journals()
    ['bmj', 'cell', 'colorblind', 'jama', 'lancet', 'nature', 'nejm', 'pnas', 'science']
    """
    return sorted(_get_data().keys())


def list_palettes(journal: str) -> list[str]:
    """Return all palette names for a given journal.

    Parameters
    ----------
    journal : str
        Journal key (case-insensitive). Use :func:`list_journals` to see options.

    Returns
    -------
    list[str]
        Sorted list of palette names for that journal.

    Raises
    ------
    ValueError
        If the journal key is not found.

    Examples
    --------
    >>> import pubchroma as pc
    >>> pc.list_palettes("nature")
    ['light', 'main']
    """
    data = _get_data()
    key = journal.lower()
    if key not in data:
        available = ", ".join(sorted(data.keys()))
        raise ValueError(f"Journal '{journal}' not found. Available: {available}")
    return sorted(data[key]["palettes"].keys())


def get_palette(journal: str, palette: str = "main") -> dict:
    """Return full palette metadata for a journal.

    Parameters
    ----------
    journal : str
        Journal key (case-insensitive).
    palette : str, optional
        Palette name within that journal, by default ``"main"``.

    Returns
    -------
    dict
        Dictionary with keys: ``colors``, ``colorblind_safe``, ``description``, ``type``.

    Raises
    ------
    ValueError
        If the journal or palette is not found.

    Examples
    --------
    >>> import pubchroma as pc
    >>> p = pc.get_palette("nature")
    >>> p["colors"][:3]
    ['#E64B35', '#4DBBD5', '#00A087']
    """
    data = _get_data()
    key = journal.lower()
    if key not in data:
        available = ", ".join(sorted(data.keys()))
        raise ValueError(f"Journal '{journal}' not found. Available: {available}")
    palettes = data[key]["palettes"]
    pal = palette.lower()
    if pal not in palettes:
        available = ", ".join(sorted(palettes.keys()))
        raise ValueError(f"Palette '{palette}' not found for '{journal}'. Available: {available}")
    return palettes[pal]


def get_colors(
    journal: str,
    palette: str = "main",
    n: Optional[int] = None,
    colorblind_only: bool = False,
) -> list[str]:
    """Return hex color codes for a journal palette.

    Parameters
    ----------
    journal : str
        Journal key (case-insensitive).
    palette : str, optional
        Palette name, by default ``"main"``.
    n : int, optional
        Number of colors to return. If None, returns all colors.
        If ``n`` exceeds the palette length, colors are cycled.
    colorblind_only : bool, optional
        If True, raise an error when the palette is not colorblind-safe.

    Returns
    -------
    list[str]
        List of hex color strings (e.g. ``['#E64B35', '#4DBBD5', ...]``).

    Raises
    ------
    ValueError
        If ``colorblind_only=True`` and the palette is not colorblind-safe,
        or if the journal/palette is not found.

    Examples
    --------
    >>> import pubchroma as pc
    >>> pc.get_colors("nature", n=3)
    ['#E64B35', '#4DBBD5', '#00A087']
    >>> pc.get_colors("colorblind", "okabe_ito", colorblind_only=True)
    ['#E69F00', '#56B4E9', '#009E73', '#F0E442', '#0072B2', '#D55E00', '#CC79A7', '#000000']
    """
    pal = get_palette(journal, palette)

    if colorblind_only and not pal["colorblind_safe"]:
        raise ValueError(
            f"Palette '{palette}' for '{journal}' is not colorblind-safe. "
            "Use colorblind_only=False or choose a colorblind-safe palette."
        )

    colors = pal["colors"]
    if n is None:
        return list(colors)

    if n <= 0:
        raise ValueError(f"n must be a positive integer, got {n}")

    if n <= len(colors):
        return colors[:n]

    # Cycle colors if n > palette length
    cycles = (n // len(colors)) + 1
    return (colors * cycles)[:n]
